rust/lance-index/src/vector/bq/builder.rs

```diff
@@ -0,0 +1,412 @@
+            _ => unimplemented!("RabitQ does not support data type: {:?}", T::FLOAT_TYPE),
+        };
+
+        let metadata = RabitQuantizationMetadata {
+            rotate_mat: Some(rotate_mat),
+            rotate_mat_position: 0,
+            num_bits,
+            packed: false,
+        };
+        Self { metadata }
+    }
+
+    pub fn num_bits(&self) -> u8 {
+        self.metadata.num_bits
+    }
+
+    #[inline]
+    fn rotate_mat_flat<T: ArrowFloatType>(&self) -> &[T::Native] {
+        let rotate_mat = self.metadata.rotate_mat.as_ref().unwrap();
+        rotate_mat
+            .values()
+            .as_any()
+            .downcast_ref::<T::ArrayType>()
+            .unwrap()
+            .as_slice()
+    }
+
+    #[inline]
+    fn rotate_mat<T: ArrowFloatType>(&'_ self) -> ndarray::ArrayView2<'_, T::Native> {
+        let code_dim = self.code_dim();
+        ndarray::ArrayView2::from_shape((code_dim, code_dim), self.rotate_mat_flat::<T>()).unwrap()
+    }
+
+    pub fn dim(&self) -> usize {
+        self.code_dim() / self.metadata.num_bits as usize
+    }
+
+    // compute the dot product of v_q * v_r
+    pub fn codes_res_dot_dists<T: ArrowFloatType>(
+        &self,
+        residual_vectors: &FixedSizeListArray,
+    ) -> Result<Vec<f32>>
+    where
+        T::Native: AsPrimitive<f32>,
+    {
+        let dim = self.dim();
+        if residual_vectors.value_length() as usize != dim {
+            return Err(Error::invalid_input(
+                format!(
+                    "Vector dimension mismatch: {} != {}",
+                    residual_vectors.value_length(),
+                    dim
+                ),
+                location!(),
+            ));
+        }
+
+        // convert the vector to a dxN matrix
+        let vec_mat = ndarray::ArrayView2::from_shape(
+            (residual_vectors.len(), dim),
+            residual_vectors
+                .values()
+                .as_any()
+                .downcast_ref::<T::ArrayType>()
+                .unwrap()
+                .as_slice(),
+        )
+        .map_err(|e| Error::invalid_input(e.to_string(), location!()))?;
+        let vec_mat = vec_mat.t();
+
+        let rotate_mat = self.rotate_mat::<T>();
+        // slice to (code_dim, dim)
+        let rotate_mat = rotate_mat.slice(s![.., 0..dim]);
+        let rotated_vectors = rotate_mat.dot(&vec_mat);
+        let sqrt_dim = (dim as f32 * self.metadata.num_bits as f32).sqrt();
+        let norm_dists = rotated_vectors.mapv(|v| v.as_().abs()).sum_axis(Axis(0)) / sqrt_dim;
+        debug_assert_eq!(norm_dists.len(), residual_vectors.len());
+        Ok(norm_dists.to_vec())
+    }
+
+    fn transform<T: ArrowFloatType>(
+        &self,
+        residual_vectors: &FixedSizeListArray,
+    ) -> Result<ArrayRef>
+    where
+        T::Native: AsPrimitive<f32>,
+    {
+        // we don't need to normalize the residual vectors,
+        // because the signal of P^{-1} x v_r is the same as P^{-1} x v_r / ||v_r||
```

Comment:
How do we know the signal is the same?

Reply:
Ah, I see! Nit: can you change the comment to say "sign" instead of "signal"? I think that is the more common term, and "signal" doesn't appear in the paper.